### PR TITLE
Game engine integration

### PR DESCRIPTION
**Pull request by CSimbulan**
Integration of the game engine with the front end. The game is pretty much playable now.

When the players enter the game screen, the blue master is immediately given a modal requesting a clue word and the max amount of guesses. The socket takes this info, updates the game engine and sends a new version and the blue guessers can start picking a card. If they click on a card it will get added to an array of "voted" cards. The engine checks if the number of votes is equal to number of guessers for that team, and after the final vote it counts for the most voted card (if it's a tie then it's whichever got picked first out of the winners) and that card gets checked in the game engine. A point will be given correcty and the turn may continue or it could go over to the next team depending on the situation. 

Once the blue team is out of guesses or they guess wrong, the red spymaster gets a modal like the blue spymaster did, and the turn plays out similiarly, and this will alternate until a win condition is met. 

There is currently no game over screen, just console logs to note a game ending. There is also no option to pass right now, I need to figure out where to put the button (probably on the chat side of the board).

Also I'm still not 100% sure what to do with the sockets because what I have here is the only it won't eventually crash unless we redesign the whole socket structure (can talk about this in standup)

Also for the card flipping library, I looked into MUI if there was a component to do this easily, the transition component looks like it could possibly do it but the documentation states "(Fade, Grow, Zoom, Slide)" so I think using the library I added easier for now.

Some screenshots (I removed the chat for now to give a better view since it the real thing is designed for full screen):

Initial board state:

After successful blue team guess and start of red team's turn:

Red team guessing phase:

### CODE REVIEW

File: client/src/components/UserPrompt.js [hunk 1]
```diff
@@ -49,10 +57,28 @@ const UserPrompt = () => {
...
+              matchState.turnPhase === "guess" &&
+              matchState.redGuessers.findIndex(
+                (player) => player.email === email
+              ) !== -1) ||
+            (matchState.turn === "blue" &&
+              matchState.turnPhase === "clue" &&
+              matchState.blueSpymaster.email === email) ||
+            (matchState.turn === "red" &&
+              matchState.turnPhase === "clue" &&
+              matchState.redSpymaster.email === email)) && snackbarOpen
```
Comment (uzher-code) [inline client/src/components/UserPrompt.js]: seems redundant with what is happening in useEffect, since these checks are already happening there.

Comment (CSimbulan) [reply client/src/components/UserPrompt.js]: yes you're right I'm checking for the user role again but I can just use the state, I changed it now

File: client/src/components/GameBoard.js [hunk 2]
```diff
@@ -32,27 +35,79 @@ const useStyles = makeStyles({
...
     const classes = useStyles();
     const { email } = useUserState();
-    const { matchState } = useContext(MatchContext);
+    const { matchState, setMatchState } = useContext(MatchContext);
+    const socketRef = useRef();
+    const { matchId } = useParams();
+    const [selected, setSelected] = useState({});
     // TODO: integrate with backend.
     //function onCardClick() {
     //}
```
Comment (sina-jamshidi) [inline client/src/components/GameBoard.js]: this can be removed with the latest changes I think

File: client/src/components/GameBoard.js [hunk 3]
```diff
@@ -32,27 +35,79 @@ const useStyles = makeStyles({
...
+            socketRef.current.emit("get-game-engine", { matchId });
+        }
+        socketRef.current.on("update-game-engine-" + matchId, (game) => {
+            setMatchState(game);
+            if (matchState.votes === {}) {
+                setSelected({});
+            }
+        });
+        return () => socketRef.current.disconnect();
+    }, [matchState, setMatchState, matchId]);
```
Comment (sina-jamshidi) [inline client/src/components/GameBoard.js]: Shouldn't the dependency array be empty? This seems like something we only run once

File: client/src/components/GameBoard.js [hunk 4]
```diff
@@ -32,27 +35,79 @@ const useStyles = makeStyles({
...
+
+    const handleVote = (word, row, column) => {
+        console.log(word, row, column);
+        setSelected({ row, column });
+        socketRef.current.emit("set-vote", { matchId, word, row, column, email });
+        return "x";
```
Comment (sina-jamshidi) [inline client/src/components/GameBoard.js]: why are we returning `x`?

**Review by sina-jamshidi — APPROVED** [review]
looks good to merge

File: client/src/components/GameBoard.js (hunk 4 above)
Comment (CSimbulan) [reply client/src/components/GameBoard.js]: Was using it to test something before but forgot to remove it too

File: client/src/components/GameBoard.js (hunk 3 above)
Comment (CSimbulan) [reply client/src/components/GameBoard.js]: Yes you're right

**Pull request merged**